File: 720. Longest Word in Dictionary/720.go
package leetcode

import (
	"math"
	"sort"
)

func longestWord(words []string) string {
	myLen := 1
	myDict := make(map[string]int, 0)
	nextList := make([]string, 0)

	for _,v := range words{
		if len(v) == myLen{
			myDict[v] = 1
		}else{
			nextList = append(nextList, v)
		}
	}
	myLen ++
	words = nextList

	for len(nextList) != 0{
		nextList := make([]string, 0)
		tmp := make(map[string]int, 0)
		for _,v := range words{
			if len(v) == myLen{
				_, ok := myDict[v[:myLen-1]]
				if ok{
					tmp[v] = 1
				}
			}else{
				nextList = append(nextList, v)
			}
		}
		if len(tmp) == 0{
			break
		}
		myLen ++
		words = nextList
		myDict = tmp
	}

	res := "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"
	for i := range myDict{
		if res > i{
			res = i
		}
	}
	return res
}

func longestWord1(words []string) string {
    if len(words) < 2 { return "" }
	tmp := []string{}
	max := math.MinInt32
	sort.Slice(words, func(i, j int) bool {
		return len(words[i]) < len(words[j])
	})
	for i := len(words) - 1; i > 0; i-- {
		count := 1
		l := len(words[i])
		for j := i - 1; j >= 0; j-- {
			if words[i][:l-count] == words[j] {
				count++
			}
			if count == l {
				if len(words[i]) > max {
					max = len(words[i])
				}
				if len(words[i]) == max {
					tmp = append(tmp, words[i])
				}
			}
		}
	}
    if len(tmp) == 1 { return tmp[0] }
	res := tmp[0]
	for _, v := range tmp {
		if v < res {
			res = v
		}
	}
	return res
}

File: 720. Longest Word in Dictionary/_720.py
class Solution:
    def longestWord(self, words) -> str:
        words.sort(key = lambda i:len(i)) 
        mylen = 1

        tmp = {}
        end = 0
        for i,v in enumerate(words):
            if len(v) == mylen:
                tmp[v] = 1
                end = i
        words = words[end+1:]
        end =0
        mylen += 1

        while len(words) != 0:
            tmp1 ={}
            for i,v in enumerate(words):
                if len(v) == mylen:
                    dictv = tmp.get(v[:mylen-1], 0)
                    if dictv != 0:
                        tmp1[v] = 1
                    end = i
            if len(tmp1) == 0:
                break
            else:
                tmp = tmp1
                words = words[end+1:]
                end =0
                mylen += 1

        res = [i for i in tmp]
        res.sort()
        return res[0]


    def longestWord1(self, words):
        words.sort()
        words_set, longest_word = set(['']), ''
        for word in words:
            if word[:-1] in words_set:
                words_set.add(word)
                if len(word) > len(longest_word):
                    longest_word = word
        return longest_word

    # 这个思路还是一样，但写的更python
    def longestWord2(self, words):
        valid = set([""])
        for word in sorted(words, key=lambda x: len(x)):
            if word[:-1] in valid:
                valid.add(word)
        return max(sorted(valid), key=lambda x: len(x))

a = Solution()
print(a.longestWord( ["a", "banana", "app", "appl", "ap", "apply", "apple"]))